### PR TITLE
Issues seen during concurrency testing

### PR DESCRIPTION
**Pull request by hbhanawat**
## Changes proposed in this pull request

CachedDataFrame.queryExecutionString was a function which is getting evaluated every time the CDF is being executed. Changed it to lazy val so that it is evaluated once. Same for the number of partitions of the cached RDD.  

## Patch testing

precheckin

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala [hunk 1]
```diff
@@ -88,7 +88,9 @@ class CachedDataFrame(session: SparkSession, queryExecution: QueryExecution,
   private lazy val boundEnc = exprEnc.resolveAndBind(logicalPlan.output,
     sparkSession.sessionState.analyzer)
 
-  def queryExecutionString: String = queryExecution.toString()
+  private lazy val queryExecutionString: String = queryExecution.toString()
+
+  private lazy val numPartitions: Int = cachedRDD.getNumPartitions
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Won't this cause trouble if the pruning is different for different tokens? From what I know Soubhik had handled that by explicitly clearing the cached partitions inside cachedRDD and re-calculating in every execution.

File: core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala [hunk 2]
```diff
@@ -88,7 +88,9 @@ class CachedDataFrame(session: SparkSession, queryExecution: QueryExecution,
   private lazy val boundEnc = exprEnc.resolveAndBind(logicalPlan.output,
     sparkSession.sessionState.analyzer)
 
-  def queryExecutionString: String = queryExecution.toString()
+  private lazy val queryExecutionString: String = queryExecution.toString()
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Need to be able to generate appropriate plans for different tokens as discussed. @kneeraj can you track that issue?

File: core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala [hunk 3]
```diff
@@ -146,7 +149,7 @@ class CachedDataFrame(session: SparkSession, queryExecution: QueryExecution,
       sessionState.conf.activeSchedulerPool
 
     // Check if it is pruned query, execute it automatically on the low latency pool
-    if ((cachedRDD ne null) && cachedRDD.getNumPartitions <= 2 /* some small number */ &&
+    if ((cachedRDD ne null) && isLowLatencyQuery &&
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Shouldn't the (cachedRDD ne null) check also go into isLowLatencyQuery flag above? I think there are some cases where cachedRDD can be null like insert/update/delete in which case isLowLatencyQuery should be false like here.

**Comment by hbhanawat** [comment]
Create SNAP-1931, SNAP-1932

**Pull request merged**